util/circular_buffer.c

```diff
@@ -0,0 +1,246 @@
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include "circular_buffer.h"
+#include "error.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @struct circular_buffer
+ * @brief Circular buffer descriptor
+ */
+struct circular_buffer {
+	/** Size of the buffer in bytes */
+	uint32_t	size;
+	/** Number of user elements in the buffer */
+	uint32_t	nb_elem;
+	/** Size of an user element */
+	uint32_t	elem_size;
+	/** Address to the buffer */
+	int8_t		*buff;
+	/** Pointer to reading location */
+	int8_t		*read_ptr;
+	/** Pointer to writing location */
+	int8_t		*write_ptr;
+	/** Set when the buffer is full */
+	bool		full;
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+int32_t circular_buffer_init(circular_buffer_t *desc, uint32_t nb_elements,
+			     uint32_t element_size)
+{
+	struct circular_buffer	*ldesc;
+	uint32_t		size;
+
+	if (!desc || !element_size || !nb_elements)
+		return FAILURE;
+
+	size = nb_elements * element_size;
+	if (size < nb_elements || size < element_size)
+		return FAILURE; //Integer overflow
```

Comment:
I am not convinced that this catches all scenarios. Somehow I have a gut feeling there is a counterexample that may get undetected by this check.
Worth checking: https://stackoverflow.com/questions/1815367/catch-and-compute-overflow-during-multiplication-of-two-large-integers

Reply:
nb_elemnts and elemnt_size are unsigned so I think there is no counterexample. If one of them is 0 is checked in the previous if.